rotkehlchen/chain/ethereum/adex/adex.py

```diff
@@ -242,36 +286,38 @@ def _get_staking_history(
             adx_staking_details = []
             for adx_staking_balance in adx_staking_balances:
                 # NB: currently it only returns staking details for Tom pool
-                if adx_staking_balance.pool_id == TOM_POOL_ID and tom_pool_incentive is not None:
-                    adx_total_profit = Balance()
-                    dai_total_profit = Balance()
-                    # Add claimed amounts and their historical usd value
-                    for event in address_staking_events[address]:
-                        if isinstance(event, ChannelWithdraw):
-                            if event.token == A_ADX:
-                                adx_total_profit += event.value
-                            elif event.token == A_DAI:
-                                dai_total_profit += event.value
-
-                    # Add unclaimed amounts and their current usd value
-                    adx_total_profit += adx_staking_balance.adx_unclaimed_balance
-                    dai_total_profit += adx_staking_balance.dai_unclaimed_balance
-
-                    pool_staking_detail = ADXStakingDetail(
-                        contract_address=adx_staking_balance.contract_address,
-                        pool_id=adx_staking_balance.pool_id,
-                        pool_name=adx_staking_balance.pool_name,
-                        total_staked_amount=(
-                            tom_pool_incentive.total_staked_amount / ADX_AMOUNT_MANTISSA
-                        ),
-                        apr=tom_pool_incentive.apr,
-                        adx_balance=adx_staking_balance.adx_balance,
-                        adx_unclaimed_balance=adx_staking_balance.adx_unclaimed_balance,
-                        dai_unclaimed_balance=adx_staking_balance.dai_unclaimed_balance,
-                        adx_profit_loss=adx_total_profit,
-                        dai_profit_loss=dai_total_profit,
-                    )
-                    adx_staking_details.append(pool_staking_detail)
+                if adx_staking_balance.pool_id != TOM_POOL_ID:
+                    continue
+
+                adx_total_profit = Balance()
+                dai_total_profit = Balance()
+                # Add claimed amounts and their historical usd value
+                for event in address_staking_events[address]:
+                    if isinstance(event, ChannelWithdraw):
+                        if event.token == A_ADX:
+                            adx_total_profit += event.value
+                        elif event.token == A_DAI:
+                            dai_total_profit += event.value
+
+                # Add unclaimed amounts and their current usd value
+                adx_total_profit += adx_staking_balance.adx_unclaimed_balance
```

Comment:
In the case of to timestamp not being "now" I guess you have no way of knowing if the unclaimed balance exists or even how much it is, so in that case you would need to not take it into account.

Such case would only come in from the PnL report. So imagine I ask for the PnL of Adex in 2020.

Another thing required there would be the events in that timestamp from which you calculated the profit from. I guess they are already available in `address_staking_events[address]` so they can probably just be included in `ADXStakingDetail`?

This way when we call them in the pnl report [here](https://github.com/rotki/rotki/blob/54ee785593598e04a922d8b0309a53f695aff898/rotkehlchen/history/events.py#L339-L362) we can get both amounts and events for the given period.

I am actively working on that "DefiEvents" part of the PnL report now. It needs more work and thinking on how to handle all different kind of events in a given period for each protocol, and how to show them in the exported CSV for the user.

I think we are gonna iteratively improve on it

Reply:
Okay, this PnL definitely is challenging.
I was thinking the safest way to make it work would be to introduce a new variable `is_pnl_report: boolean` in `get_events_history`. When calling `get_events_history` for the PnL report the following happens:
- **We do only rely on events stored in the DB** (this assumption may be wrong). The db ChannelWithdraw events already have associated a token (ADX or DAI). Therefore there is no need to call the fee-rewards API.
- `_calculate_staking_balances()` returns `ADXStakingBalance`, which has the ADX amount for the given timestamp range, and both unclaimed ADX and unclaimed DAI are set to 0.
- `_get_staking_history()` will work only with the events for the given timestamp range. `ADXStakingDetail` will have set to 0 the pool stats and unclaimed ADX/DAI. Profit and loss amounts won't take into account the unclaimed amounts (as they are 0).

Would this work for the PnL? I understand that the assumption of only relying on db events may be wrong if they were never requested in the first place for the given timestamp range. I'm just trying to do not rely on the fee-rewards API.

Reply:
So all valid questions. Let me get to them one by one. Please note that whatever hold true for adex here will hold true for every single other module in rotki. And similar challenges are posed by other modules.

> We do only rely on events stored in the DB (this assumption may be wrong).

There is no need to avoid calling external apis there. We already do. But it's **much better** if we don't. So something like the following is what we try to do in the history call. The history calls of all modules are called both by the UI but also by the PnL report. For the PnL:

1. If the user has never ever looked at their adex history this call would need to call the full history and then save it in the DB
2. If the user has already looked at their adex history it should already be in the DB.
3. If the timerange events are all in the DB (say all of 2020) and no more is required it should try to figure stuff out only by the DB events. **this is the preferred scenario**
4. If only part of the range is in the DB then the rest should be requested by whatever means exist ... for example on chain event calls, graph queries, etc

What we always try to make happen is (3). We also have these periodic backend queries which run on "idle time" which try to call history for all modules, trades, exchanges etc. So the idea is to get all history, store it locally so that when it's needed we got it. So that we end up in (3) as often as possible.


> _calculate_staking_balances()
> _get_staking_history()

I don't know of the exact details per function but the idea is to know how much the user has gained in a given time range. And if possible (atm it's not possible in all modules -- which is something we should fix) show the relevant events. In DeFi events, in another PR I am adding a transaction hashes attribute, so that the user can at least look them up when looking at the CSV. As things progress we may end up exposing more stuff in the CSV, so having the events available is useful.

Reply:
Resolving this as in the PNL report it will be zero.